File: DSA Unit 05/Lec 31/Q3.cpp
Single File Programming Question
Problem Statement



﻿Hema, a software developer, is working on a problem where she needs to find the length of the longest subarray with distinct elements in a given array. The subarray should not have any repeated elements, and Hema needs to implement a solution using a sliding window technique. Help her implement this task.

Input format :
The first line of input consists of an integer n, representing the number of elements in the array.

The second line of input consists of n integers, representing the elements of the array.

Output format :
The first line of output prints: "Longest subarray length: " followed by the length of the longest subarray with distinct elements.



Refer to the sample output for formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

1 ≤ n ≤ 1000
The elements in the array are integers.
MAX_SIZE = 1000
Sample test cases :
Input 1 :
7
1 2 3 4 5 6 7
Output 1 :
Longest subarray length: 7
Input 2 :
8
1 2 2 3 4 5 6 7
Output 2 :
Longest subarray length: 6



#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;

    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    unordered_set<int> window;
    int left = 0, longest = 0;

    for (int right = 0; right < n; right++) {
        while (window.find(arr[right]) != window.end()) {
            window.erase(arr[left]);
            left++;
        }
        window.insert(arr[right]);
        longest = max(longest, right - left + 1);
    }

    cout << "Longest subarray length: " << longest;
    return 0;
}

File: Java/Unit 03/Lec 19/Q3.java
Single File Programming Question
Problem Statement



Eric operates a taxi service and charges passengers based on the distance traveled and the waiting time. The total fare is calculated using a base fare, cost per mile, and cost per minute of wait time.



Implement this calculation using an abstract class `FareCalculator` and an interface `FareDetails`. The abstract class should define an abstract method `calculateFare`, while the interface should declare constants for base fare, cost per mile, and cost per minute.



Formula: 

Total Fare = base fare + (distance * cost per mile) + (waiting time * cost per minute)



Eric's Standard Fee Structure:

Base fare = 3.50
Cost per mile = 2.50
Cost per minute = 0.20
Input format :
The first line of the input consists of a double value representing the distance traveled in miles.

The second line consists of an integer representing the waiting time in minutes (a non-negative value).

Output format :
The output prints "Total Fare: " followed by a double value representing the total fare, rounded to two decimal places.



Refer to the sample output formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

0.0 ≤ distance travelled ≤ 50.0

1 ≤ waiting time ≤ 60

Sample test cases :
Input 1 :
7.36
6
Output 1 :
Total Fare: 23.10
Input 2 :
13.25
5
Output 2 :
Total Fare: 37.63
Whitelist
Set 1:
abstract
interface
implements


import java.util.*;

// Interface jo constant values (rates) hold karti hai
interface FareDetails {
    double BASE_FARE = 3.50;
    double COST_PER_MILE = 2.50;
    double COST_PER_MINUTE = 0.20;
}

// Abstract class jo method structure define karti hai
abstract class FareCalculator {
    public abstract double calculateFare(double distance, int waitingTime);
}

// Concrete class jo abstract class ko extend aur interface ko implement karti hai
class TaxiFareCalculator extends FareCalculator implements FareDetails {
    
    @Override
    public double calculateFare(double distance, int waitingTime) {
        // Formula: Total Fare = base fare + (distance * cost per mile) + (waiting time * cost per minute)
        double fare = BASE_FARE + (distance * COST_PER_MILE) + (waitingTime * COST_PER_MINUTE);
        return fare;
    }
}

// Footer Snippet (Platform dwara pehle se diya gaya hai)
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        // Distance aur Waiting time input lena
        double distance = sc.nextDouble();
        int waitingTime = sc.nextInt();
        
        // TaxiFareCalculator ka object banana
        FareCalculator calculator = new TaxiFareCalculator();
        double totalFare = calculator.calculateFare(distance, waitingTime);
        
        // Output ko 2 decimal places tak format karna
        System.out.printf("Total Fare: %.2f", totalFare);
    }
}
